### PR TITLE
Add tests for IrSyntax and fix #739

### LINKED ISSUE #739: ExtensionProof Error
This issue describes the compiler error in [this](https://github.com/arrow-kt/arrow-meta/blob/24460a6f2f7d0d53aa3770e04e1a9b5d48c0295a/compiler-plugin/src/test/kotlin/arrow/meta/plugins/typeclasses/TypeClassesTest.kt#L38) Test and generalizes to all insertions of extension proofs. 
Finishing with 

<details>
<summary>Compiler Error</summary>
error found: java.lang.AssertionError: Substituted containing declaration: public abstract fun empty(): kotlin.String defined in arrowx.Monoid[DeserializedSimpleFunctionDescriptor@7fc3b3c0]
for descriptor: <this> defined in arrowx.Monoid.empty[ReceiverParameterDescriptorImpl@43d36b9c]
	at org.jetbrains.kotlin.ir.symbols.impl.IrBindableSymbolBase.<init>(IrPrivateSymbolBase.kt:43)
	at org.jetbrains.kotlin.ir.symbols.impl.IrValueParameterSymbolImpl.<init>(IrPrivateSymbolBase.kt:111)
	at org.jetbrains.kotlin.ir.declarations.impl.IrValueParameterImpl.<init>(IrValueParameterImpl.kt:57)
	at org.jetbrains.kotlin.ir.declarations.lazy.IrLazyDeclarationBase.generateReceiverParameterStub(IrLazyDeclarationBase.kt:30)
	at org.jetbrains.kotlin.ir.declarations.lazy.IrLazyFunctionBase$dispatchReceiverParameter$2.invoke(IrLazyFunctionBase.kt:42)
	at org.jetbrains.kotlin.ir.declarations.lazy.IrLazyFunctionBase$dispatchReceiverParameter$2.invoke(IrLazyFunctionBase.kt:21)
	at org.jetbrains.kotlin.ir.declarations.lazy.UnsafeLazyVar$value$1.invoke(lazyUtil.kt:21)
	at org.jetbrains.kotlin.ir.declarations.lazy.UnsafeLazyVar$value$1.invoke(lazyUtil.kt:13)
	at org.jetbrains.kotlin.ir.declarations.StageController$DefaultImpls.withInitialIr(PersistentApi.kt:27)
	at org.jetbrains.kotlin.ir.declarations.PersistentApiKt$stageController$1.withInitialIr(PersistentApi.kt:13)
	at org.jetbrains.kotlin.ir.declarations.lazy.UnsafeLazyVar.getValue(lazyUtil.kt:38)
	at org.jetbrains.kotlin.ir.declarations.lazy.UnsafeLazyVar.getValue(lazyUtil.kt:31)
	at org.jetbrains.kotlin.ir.declarations.lazy.IrLazyFunctionBase.getDispatchReceiverParameter(IrLazyFunctionBase.kt)
	at org.jetbrains.kotlin.backend.common.CheckIrElementVisitor.checkFunction(CheckIrElementVisitor.kt:87)
	at org.jetbrains.kotlin.backend.common.CheckIrElementVisitor.visitCall(CheckIrElementVisitor.kt:168)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitCall(IrElementVisitorVoid.kt:165)
	at org.jetbrains.kotlin.backend.common.CheckIrElementVisitor.visitCall(CheckIrElementVisitor.kt:34)
	at org.jetbrains.kotlin.backend.common.CheckIrElementVisitor.visitCall(CheckIrElementVisitor.kt:34)
	at org.jetbrains.kotlin.ir.expressions.impl.IrCallImpl.accept(IrCallImpl.kt:78)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoidKt.acceptVoid(IrElementVisitorVoid.kt:271)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitElement(IrValidator.kt:86)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitExpression(IrElementVisitorVoid.kt:104)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitExpression(IrValidator.kt:60)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitMemberAccess(IrElementVisitorVoid.kt:158)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitMemberAccess(IrValidator.kt:60)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitFunctionAccess(IrElementVisitorVoid.kt:161)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitFunctionAccess(IrValidator.kt:60)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitCall(IrElementVisitorVoid.kt:164)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitCall(IrValidator.kt:60)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitCall(IrElementVisitorVoid.kt:165)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitCall(IrValidator.kt:60)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitCall(IrValidator.kt:60)
	at org.jetbrains.kotlin.ir.expressions.impl.IrCallImpl.accept(IrCallImpl.kt:78)
	at org.jetbrains.kotlin.ir.expressions.impl.IrExpressionBodyImpl.acceptChildren(IrExpressionBodyImpl.kt:50)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoidKt.acceptChildrenVoid(IrElementVisitorVoid.kt:275)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitElement(IrValidator.kt:87)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitBody(IrElementVisitorVoid.kt:86)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitBody(IrValidator.kt:60)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitExpressionBody(IrElementVisitorVoid.kt:89)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitExpressionBody(IrValidator.kt:60)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitExpressionBody(IrElementVisitorVoid.kt:90)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitExpressionBody(IrValidator.kt:60)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitExpressionBody(IrValidator.kt:60)
	at org.jetbrains.kotlin.ir.expressions.impl.IrExpressionBodyImpl.accept(IrExpressionBodyImpl.kt:47)
	at org.jetbrains.kotlin.ir.declarations.impl.IrFieldImpl.acceptChildren(IrFieldImpl.kt:122)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoidKt.acceptChildrenVoid(IrElementVisitorVoid.kt:275)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitElement(IrValidator.kt:87)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitDeclaration(IrElementVisitorVoid.kt:40)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitDeclaration(IrValidator.kt:60)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitField(IrElementVisitorVoid.kt:61)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitField(IrValidator.kt:60)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitField(IrElementVisitorVoid.kt:62)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitField(IrValidator.kt:60)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitField(IrValidator.kt:60)
	at org.jetbrains.kotlin.ir.declarations.impl.IrFieldImpl.accept(IrFieldImpl.kt:118)
	at org.jetbrains.kotlin.ir.declarations.impl.IrPropertyCommonImpl.acceptChildren(IrPropertyImpl.kt:90)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoidKt.acceptChildrenVoid(IrElementVisitorVoid.kt:275)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitElement(IrValidator.kt:87)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitDeclaration(IrElementVisitorVoid.kt:40)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitDeclaration(IrValidator.kt:60)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitProperty(IrElementVisitorVoid.kt:58)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitProperty(IrValidator.kt:60)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitProperty(IrElementVisitorVoid.kt:59)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitProperty(IrValidator.kt:60)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitProperty(IrValidator.kt:60)
	at org.jetbrains.kotlin.ir.declarations.impl.IrPropertyCommonImpl.accept(IrPropertyImpl.kt:86)
	at org.jetbrains.kotlin.ir.declarations.impl.IrFileImpl.acceptChildren(IrFileImpl.kt:59)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoidKt.acceptChildrenVoid(IrElementVisitorVoid.kt:275)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitElement(IrValidator.kt:87)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitPackageFragment(IrElementVisitorVoid.kt:30)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitPackageFragment(IrValidator.kt:60)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitFile(IrElementVisitorVoid.kt:37)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitFile(IrValidator.kt:67)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitFile(IrElementVisitorVoid.kt:38)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitFile(IrValidator.kt:60)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitFile(IrValidator.kt:60)
	at org.jetbrains.kotlin.ir.declarations.impl.IrFileImpl.accept(IrFileImpl.kt:56)
	at org.jetbrains.kotlin.ir.declarations.impl.IrModuleFragmentImpl.acceptChildren(IrModuleFragmentImpl.kt:44)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoidKt.acceptChildrenVoid(IrElementVisitorVoid.kt:275)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitElement(IrValidator.kt:87)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitModuleFragment(IrElementVisitorVoid.kt:27)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitModuleFragment(IrValidator.kt:60)
	at org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid$DefaultImpls.visitModuleFragment(IrElementVisitorVoid.kt:28)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitModuleFragment(IrValidator.kt:60)
	at org.jetbrains.kotlin.backend.common.IrValidator.visitModuleFragment(IrValidator.kt:60)
	at org.jetbrains.kotlin.ir.declarations.impl.IrModuleFragmentImpl.accept(IrModuleFragmentImpl.kt:41)
	at org.jetbrains.kotlin.backend.common.phaser.DumperVerifierKt.validationCallback(DumperVerifier.kt:134)
	at org.jetbrains.kotlin.backend.jvm.JvmLowerKt$validateIrBeforeLowering$1.invoke(JvmLower.kt:49)
	at org.jetbrains.kotlin.backend.jvm.JvmLowerKt$validateIrBeforeLowering$1.invoke(JvmLower.kt)
	at org.jetbrains.kotlin.backend.common.phaser.PhaseBuildersKt$makeCustomPhase$1.invoke(PhaseBuilders.kt:121)
	at org.jetbrains.kotlin.backend.common.phaser.PhaseBuildersKt$makeCustomPhase$1.invoke(PhaseBuilders.kt:114)
	at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper$runBody$1.invoke(CompilerPhase.kt:128)
	at org.jetbrains.kotlin.backend.common.phaser.CompilerPhaseKt.downlevel(CompilerPhase.kt:24)
	at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.runBody(CompilerPhase.kt:127)
	at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.invoke(CompilerPhase.kt:105)
	at org.jetbrains.kotlin.backend.common.phaser.CompositePhase.invoke(PhaseBuilders.kt:23)
	at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper$runBody$1.invoke(CompilerPhase.kt:128)
	at org.jetbrains.kotlin.backend.common.phaser.CompilerPhaseKt.downlevel(CompilerPhase.kt:24)
	at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.runBody(CompilerPhase.kt:127)
	at org.jetbrains.kotlin.backend.common.phaser.AbstractNamedPhaseWrapper.invoke(CompilerPhase.kt:105)
	at org.jetbrains.kotlin.backend.common.phaser.CompilerPhaseKt.invokeToplevel(CompilerPhase.kt:42)
	at org.jetbrains.kotlin.backend.jvm.JvmLower.lower(JvmLower.kt:393)
	at org.jetbrains.kotlin.backend.jvm.JvmBackendFacade.doGenerateFilesInternal$backend_jvm(JvmBackendFacade.kt:124)
	at org.jetbrains.kotlin.backend.jvm.JvmBackendFacade.doGenerateFilesInternal$backend_jvm$default(JvmBackendFacade.kt:108)
	at org.jetbrains.kotlin.backend.jvm.JvmBackendFacade.doGenerateFiles(JvmBackendFacade.kt:95)
	at org.jetbrains.kotlin.backend.jvm.JvmIrCodegenFactory.generateModule(JvmIrCodegenFactory.kt:40)
	at org.jetbrains.kotlin.codegen.KotlinCodegenFacade.compileCorrectFiles(KotlinCodegenFacade.java:35)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.generate(KotlinToJVMBytecodeCompiler.kt:616)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.compileModules$cli(KotlinToJVMBytecodeCompiler.kt:203)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:164)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:51)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:86)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:44)
	at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:98)
	at com.tschuchort.compiletesting.KotlinCompilation.compileKotlin(KotlinCompilation.kt:637)
	at com.tschuchort.compiletesting.KotlinCompilation.compile(KotlinCompilation.kt:810)
	at arrow.meta.plugin.testing.CompilationKt.compile(Compilation.kt:33)
	at arrow.meta.plugin.testing.CompilationAssertionsKt$interpreter$1.invoke(CompilationAssertions.kt:108)
	at arrow.meta.plugin.testing.CompilationAssertionsKt$interpreter$1.invoke(CompilationAssertions.kt)
	at arrow.meta.plugin.testing.CompilerTest.run$meta_test(CompilerTestDSL.kt:55)
	at arrow.meta.plugin.testing.CompilationAssertionsKt.assertThis(CompilationAssertions.kt:61)
	at arrow.meta.plugins.typeclasses.TypeClassesTest.polymorphic constrain(TypeClassesTest.kt:52)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:119)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.lang.Thread.run(Thread.java:748)
</details>

The critical line we inspect this snippet is the following:

```kotlin
val result = String.empty() 
// where String.empty() is synthetically reduced from String.Companion.monoid().empty()
```

`monoid()` is being injected coherently by the compiler the problem lies in [how we create](https://github.com/arrow-kt/arrow-meta/blob/fb40b7f5342158331e1682b67c6e01584efc7978/compiler-plugin/src/main/kotlin/arrow/meta/plugins/proofs/phases/resolve/scopes/ProofsSyntheticScopes.kt#L45-L65) the `SimpleFunctionDescriptor` in the SyntheticResolution phase for synthetic members, such as `empty()`, which then fails in Ir: 

Prior to 1.4.0, or earlier, the [invariant](https://github.com/JetBrains/kotlin/blob/48ae76ac286b9efd9cf08d0145f7ca1543dc282b/compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/symbols/impl/IrSymbolBase.kt#L31) for IrBindableSymbol was, among others, -- this didn't cause the any error's in master:

Imaginening descriptor being: `empty()`

Screenshot 2020-08-31 at 13 41 16

```kotlin
// invariant: assert(isOriginalDescriptor(descriptor)) -> for our snippet above passes with true 

fun isOriginalDescriptor(descriptor: DeclarationDescriptor): Boolean =
        descriptor is IrBasedDeclarationDescriptor ||
                descriptor is ValueParameterDescriptor && isOriginalDescriptor(descriptor.containingDeclaration) ||
                descriptor == descriptor.original
```

which changed to the [following](https://github.com/JetBrains/kotlin/blob/e4cd7cdbcf2269a612d571a34a82dbedcb93979b/compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/symbols/impl/IrPrivateSymbolBase.kt#L48)

```kotlin
/**
 * reduced Invariant:
 * assert(isOriginalDescriptor(descriptor) && if (descriptor !is WrappedDeclarationDescriptor<*>)
 *  descriptor.containingDeclaration == null || isOriginalDescriptor(descriptor.containingDeclaration) else true)
 * -> here the snippet fails 
 */
fun isOriginalDescriptor(descriptor: DeclarationDescriptor): Boolean =
  descriptor is WrappedDeclarationDescriptor<*> ||
    // TODO fix declaring/referencing value parameters: compute proper original descriptor
    descriptor is ValueParameterDescriptor && isOriginalDescriptor(descriptor.containingDeclaration) ||
    descriptor == descriptor.original
```

Previous attempts to solve this issue:
- modifying the members like `original`, `containing declaration`: Result - Didn't seem to work, so far.
- map the `SimpleFunctionDescriptors` , we're collecting from the ResolutionScope [here](https://github.com/arrow-kt/arrow-meta/blob/fb40b7f5342158331e1682b67c6e01584efc7978/compiler-plugin/src/main/kotlin/arrow/meta/plugins/proofs/phases/Proofs.kt#L123) to `WrappedSimpleFunctionDescriptor`, but the problem here lies in the construction of that type. It needs an IrFunction it needs to be bound to with `bind`. Here is one [example](https://github.com/JetBrains/kotlin/blob/1e96c3d21efc9a1c37d3d8c1331a7ad5f46394ec/compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/lower/CreateScriptFunctionsPhase.kt#L92). And I am not sure, how or if it is possible to retrieve the IrFunction in SyntheticResolution.

Any help is super welcome :D 
Feel free to drop any questions if something is unclear.

### PR DESCRIPTION
**Pull request by i-walker**
resolves #739
And adds a rather initial step to ensure that all DSL calls are being called. 
For now, the fix for #739 takes precedence. So failed Syntax tests are disabled until enabled.

### CODE REVIEW

File: compiler-plugin/src/test/kotlin/arrow/meta/ir/plugin/IrSyntaxPlugin.kt [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+import org.jetbrains.kotlin.ir.expressions.IrTry
+import org.jetbrains.kotlin.ir.expressions.IrTypeOperatorCall
+import org.jetbrains.kotlin.ir.expressions.IrValueAccessExpression
+import org.jetbrains.kotlin.ir.expressions.IrVararg
+import org.jetbrains.kotlin.ir.expressions.IrWhen
+import org.jetbrains.kotlin.ir.expressions.IrWhileLoop
+
+open class IrSyntaxPlugin : Meta {
+  override fun intercept(ctx: CompilerContext): List<CliPlugin> = listOf(
+    "IrSyntaxPlugin" {
```
Comment (raulraja) [inline compiler-plugin/src/test/kotlin/arrow/meta/ir/plugin/IrSyntaxPlugin.kt:84]: 👏

**Review by raulraja — APPROVED** [review]
thanks @i-walker !

File: compiler-plugin/src/test/kotlin/arrow/meta/plugins/typeclasses/TypeClassesTest.kt [hunk 2]
```diff
@@ -34,7 +34,6 @@ class TypeClassesTest {
     }
 
   @Test
-  @Disabled
```
Comment (rachelcarmena) [inline compiler-plugin/src/test/kotlin/arrow/meta/plugins/typeclasses/TypeClassesTest.kt:37]: :tada:

File: meta-test/src/main/kotlin/arrow/meta/plugin/testing/CompilerTestDSL.kt [hunk 3]
```diff
@@ -254,7 +254,10 @@ interface AssertSyntax {
    * Allows to provide several [Assert.SingleAssert].
    */
   fun allOf(vararg elements: Assert.SingleAssert): Assert =
-    if (elements.isNotEmpty()) Assert.Many(elements.asList()) else Assert.Many(emptyList())
+    if (elements.isNotEmpty()) allOf(elements.asList()) else Assert.Many(emptyList())
+
+  fun allOf(elements: List<Assert.SingleAssert>): Assert =
```
Comment (rachelcarmena) [inline meta-test/src/main/kotlin/arrow/meta/plugin/testing/CompilerTestDSL.kt:259]: Why is it necessary? The previous one accepts `vararg`

File: compiler-plugin/src/test/kotlin/arrow/meta/plugins/refinement/RefinementTests.kt [hunk 4]
```diff
@@ -73,7 +73,6 @@ class RefinementTests {
     )
   }
 
-  @Disabled
```
Comment (rachelcarmena) [inline compiler-plugin/src/test/kotlin/arrow/meta/plugins/refinement/RefinementTests.kt:76]: :tada:

File: compiler-plugin/src/test/kotlin/arrow/meta/plugins/proofs/ResolutionTests.kt [hunk 5]
```diff
@@ -44,7 +44,6 @@ class ResolutionTests {
     }
   }
 
-  @Disabled
```
Comment (rachelcarmena) [inline compiler-plugin/src/test/kotlin/arrow/meta/plugins/proofs/ResolutionTests.kt:47]: :tada:

File: compiler-plugin/src/test/kotlin/arrow/meta/ir/syntax/Utils.kt [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package arrow.meta.ir.syntax
+
+import arrow.meta.ir.plugin.IrSyntaxPlugin
+import arrow.meta.phases.codegen.ir.IrUtils
+import arrow.meta.plugin.testing.CompilerTest
+import arrow.meta.plugin.testing.assertThis
+import org.jetbrains.kotlin.cli.common.messages.CompilerMessageSeverity
+import org.jetbrains.kotlin.ir.IrElement
+
```
Comment (rachelcarmena) [inline compiler-plugin/src/test/kotlin/arrow/meta/ir/syntax/Utils.kt:9]:
```suggestion
```

**Review by rachelcarmena — APPROVED** [review]
Awesome job @i-walker :clap:

File: meta-test/src/main/kotlin/arrow/meta/plugin/testing/CompilerTestDSL.kt (hunk 3 above)
Comment (i-walker) [reply meta-test/src/main/kotlin/arrow/meta/plugin/testing/CompilerTestDSL.kt:259]: So one can do https://github.com/arrow-kt/arrow-meta/blob/7c0c3413f22e66a1a3ea17852ad7cd4540928502/compiler-plugin/src/test/kotlin/arrow/meta/ir/syntax/IrSyntaxTest.kt#L218

File: compiler-plugin/src/main/kotlin/arrow/meta/plugins/proofs/phases/ir/ProofsIrCodegen.kt [hunk 7]
```diff
@@ -173,45 +171,40 @@ class ProofsIrCodegen(
...
     val valueType = expression.dispatchReceiver?.type?.toKotlinType()
       ?: expression.extensionReceiver?.type?.toKotlinType()
       ?: (if (expression.valueArgumentsCount > 0) expression.getValueArgument(0)?.type?.toKotlinType() else null)
     val targetType =
-      (expression.symbol.owner.descriptor.dispatchReceiverParameter?.containingDeclaration as? FunctionDescriptor)?.dispatchReceiverParameter?.type
-        ?: expression.symbol.owner.descriptor.extensionReceiverParameter?.type
-        ?: expression.symbol.owner.descriptor.valueParameters.firstOrNull()?.type
+      (expression.unsubstitutedDescriptor.dispatchReceiverParameter?.containingDeclaration as? FunctionDescriptor)?.dispatchReceiverParameter?.type
+        ?: expression.unsubstitutedDescriptor.extensionReceiverParameter?.type
+        ?: expression.substitutedValueParameters.firstOrNull()?.second
```
Comment (danimontoya) [inline compiler-plugin/src/main/kotlin/arrow/meta/plugins/proofs/phases/ir/ProofsIrCodegen.kt:181]: This is what we were looking for? is it returning the resolved type??

File: compiler-plugin/src/main/kotlin/arrow/meta/plugins/proofs/phases/ir/ProofsIrCodegen.kt [hunk 8]
```diff
@@ -220,11 +213,36 @@ class ProofsIrCodegen(
               }
             }
           }
+          symbol.owner.wrapDispatcherAndExtensionReceiver(this@run)
```
Comment (danimontoya) [inline compiler-plugin/src/main/kotlin/arrow/meta/plugins/proofs/phases/ir/ProofsIrCodegen.kt:216]: 👍

**Review by danimontoya — APPROVED** [review]
Very good job buddy!! thanks @i-walker

File: compiler-plugin/src/main/kotlin/arrow/meta/plugins/proofs/phases/ir/ProofsIrCodegen.kt (hunk 7 above)
Comment (i-walker) [reply compiler-plugin/src/main/kotlin/arrow/meta/plugins/proofs/phases/ir/ProofsIrCodegen.kt:181]: that is only a shorter version for the other expression

**Pull request merged**